hibernate-core/src/main/java/org/hibernate/query/sqm/sql/BaseSqmToSqlAstConverter.java

```diff
@@ -7383,4 +7406,50 @@ public void addCteStatement(CteStatement cteStatement) {
 			cteStatements.put( cteStatement.getCteTable().getTableExpression(), cteStatement );
 		}
 	}
+
+	@Override
+	public boolean registerVisitedAssociationKey(AssociationKey associationKey) {
+		return visitedAssociationKeys.add( associationKey );
+	}
+
+	@Override
+	public void removeVisitedAssociationKey(AssociationKey associationKey) {
+		visitedAssociationKeys.remove( associationKey );
+	}
+
+	@Override
+	public boolean isAssociationKeyVisited(AssociationKey associationKey) {
+		return visitedAssociationKeys.contains( associationKey );
+	}
+
+	@Override
+	public boolean isRegisteringVisitedAssociationKeys() {
+		/*
+			 We need to avoid loops in case of eager self-referencing associations
+
+			 	E.g.
+			 	@NamedEntityGraphs({
+				@NamedEntityGraph(
+						name = "User.overview",
+						attributeNodes = { @NamedAttributeNode("name") })
+				})
+				class Sample {
+
+					@ManyToOne
+					Sample parent;
+
+					String name;
+				}
+
+				TypedQuery<Sample> query = entityManager.createQuery(
+							"select s from Sample s",
+							Sample.class
+					);
+				query.setHint( SpecHints.HINT_SPEC_LOAD_GRAPH, entityGraph );
+
+			 */
+		return creationContext.getMaximumFetchDepth() == null
+				&& ( entityGraphTraversalState != null || getLoadQueryInfluencers().hasEnabledFetchProfiles() );
```

Comment:
Isn't this also a problem if there is no entity graph or fetch profile? Maybe I'm missing something, but it seems to me that this should only depend on the nullness of maximum fetch depth i.e. use

```java
return creationContext.getMaximumFetchDepth() == null;
```

Reply:
no the problem does not occur when no load entity graph is specified,

Reply:
But why not? If we have a simple eager association, shouldn't it run into the same issue?